rustls/src/msgs/handshake.rs

```diff
@@ -2252,3 +2253,125 @@ impl HandshakeMessagePayload {
         }
     }
 }
+
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct HpkeSymmetricCipherSuite {
+    pub hpke_kdf_id: HpkeKdf,
+    pub hpke_aead_id: HpkeAead,
+}
+
+impl Default for HpkeSymmetricCipherSuite {
+    // TODO(XXX): revisit the default configuration. This is just what Cloudflare ships right now.
+    fn default() -> Self {
+        Self {
+            hpke_kdf_id: HpkeKdf::HKDF_SHA256,
+            hpke_aead_id: HpkeAead::AES_128_GCM,
+        }
+    }
+}
+
+impl Codec for HpkeSymmetricCipherSuite {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.hpke_kdf_id.encode(bytes);
+        self.hpke_aead_id.encode(bytes);
+    }
+
+    fn read(r: &mut Reader) -> Result<Self, InvalidMessage> {
+        Ok(Self {
+            hpke_kdf_id: HpkeKdf::read(r)?,
+            hpke_aead_id: HpkeAead::read(r)?,
+        })
+    }
+}
+
+impl TlsListElement for HpkeSymmetricCipherSuite {
+    const SIZE_LEN: ListLength = ListLength::U16;
+}
+
+#[derive(Clone, Debug)]
+pub struct HpkeKeyConfig {
+    pub config_id: u8,
+    pub hpke_kem_id: HpkeKem,
+    pub hpke_public_key: PayloadU16,
+    pub hpke_symmetric_cipher_suites: Vec<HpkeSymmetricCipherSuite>,
+}
+
+impl Codec for HpkeKeyConfig {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.config_id.encode(bytes);
+        self.hpke_kem_id.encode(bytes);
+        self.hpke_public_key.encode(bytes);
+        self.hpke_symmetric_cipher_suites
+            .encode(bytes);
+    }
+
+    fn read(r: &mut Reader) -> Result<Self, InvalidMessage> {
+        Ok(Self {
+            config_id: u8::read(r)?,
+            hpke_kem_id: HpkeKem::read(r)?,
+            hpke_public_key: PayloadU16::read(r)?,
+            hpke_symmetric_cipher_suites: Vec::<HpkeSymmetricCipherSuite>::read(r)?,
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct EchConfigContents {
+    pub hpke_key_config: HpkeKeyConfig,
+    pub maximum_name_length: u8,
+    pub public_name: DnsName,
+    pub extensions: PayloadU16,
+}
+
+impl Codec for EchConfigContents {
+    fn encode(&self, bytes: &mut Vec<u8>) {
+        self.hpke_key_config.encode(bytes);
+        self.maximum_name_length.encode(bytes);
+        let dns_name = &self.public_name.borrow();
+        PayloadU8::encode_slice(dns_name.as_ref().as_ref(), bytes);
+        self.extensions.encode(bytes);
+    }
+
+    fn read(r: &mut Reader) -> Result<Self, InvalidMessage> {
+        Ok(Self {
+            hpke_key_config: HpkeKeyConfig::read(r)?,
+            maximum_name_length: u8::read(r)?,
+            public_name: {
+                DnsName::try_from_ascii(PayloadU8::read(r)?.0.as_slice())
+                    .map_err(|_| InvalidMessage::InvalidServerName)?
+            },
+            extensions: PayloadU16::read(r)?,
+        })
+    }
+}
+
+#[derive(Clone, Debug)]
+pub struct EchConfig {
+    pub version: EchVersion,
+    pub contents: EchConfigContents,
```

Comment:
Unless the standards specifically define a separate `_Contents` container, maybe rewrite this to contain the contents directly?

Reply:
The standard does define the separate container in https://datatracker.ietf.org/doc/html/draft-ietf-tls-esni-17#section-4 so I think it makes sense to keep it as-is.